src/adapters/ws-pool.ts

```diff
@@ -0,0 +1,109 @@
+import { AppComponents, IWSPoolComponent } from '../types'
+
+export async function createWSPoolComponent({
+  metrics,
+  config,
+  redis
+}: Pick<AppComponents, 'metrics' | 'config' | 'redis'>): Promise<IWSPoolComponent> {
+  const maxConcurrentConnections = (await config.getNumber('MAX_CONCURRENT_CONNECTIONS')) || 100
+  const idleTimeout = (await config.getNumber('IDLE_TIMEOUT')) || 300000 // 5 minutes default
+
+  const cleanupInterval = setInterval(async () => {
+    const now = Date.now()
+    const pattern = 'ws:conn:*'
+
+    for await (const key of redis.client.scanIterator({ MATCH: pattern })) {
+      const data = await redis.client.get(key)
+      if (data) {
+        const parsed = JSON.parse(data)
+        if (now - parsed.lastActivity > idleTimeout) {
+          const id = key.replace('ws:conn:', '')
+          await releaseConnection(id)
+          metrics.increment('ws_idle_timeouts', { id })
+        }
+      }
+    }
+  }, 60000)
+
+  async function acquireConnection(id: string) {
+    const key = `ws:conn:${id}`
+
+    // First check if connection exists
+    if (await redis.client.exists(key)) {
+      throw new Error('Connection already exists')
+    }
+
+    // Then check and acquire in a transaction
+    const multi = redis.client.multi()
+
+    try {
+      const result = await multi
+        .zCard('ws:active_connections')
+        .set(key, JSON.stringify({ lastActivity: Date.now() }), {
+          NX: true,
+          EX: Math.ceil(idleTimeout / 1000)
+        })
+        .zAdd('ws:active_connections', { score: Date.now(), value: id })
+        .exec()
+
+      if (!result) {
+        throw new Error('Transaction failed')
+      }
+
+      const [totalConnections] = result as [number]
+
+      if (totalConnections && totalConnections >= maxConcurrentConnections) {
```

Comment:
Does this mean that no more than 100 users can connect/subscribe to this WS? is it safe to have it? how will we scale?

Reply:
I removed the limit